### PR TITLE
TOOLS-4263 Convert mongorestore replica-set oplog-replay tests to Go

### PR DESCRIPTION
**Pull request by autarch**
Adds `TestOplogReplayFromLocalOplogRS` in `mongorestore/oplog_replset_test.go`, which ports two JS tests into a single Go test. It records a checkpoint timestamp, performs more writes, dumps `local.oplog.rs` with a `{ts: {$gt: <checkpoint>}}` query, and replays that dump with `--oplogReplay --oplogFile`, asserting only the ops after the checkpoint are applied. This test requires a replica set, so it is gated on `testtype.ReplSetTestType` (`TOOLS_TESTING_REPLSET`) rather than the plain integration test type; that is why these two JS tests were deliberately left out of the earlier oplog edge-case conversion.

JS -> Go mapping (both map to `TestOplogReplayFromLocalOplogRS` in `mongorestore/oplog_replset_test.go`):

- `test/qa-tests/jstests/restore/oplog_replay_local_rs.js` -> `TestOplogReplayFromLocalOplogRS` (`mongorestore/oplog_replset_test.go`) - the `--oplogReplay --oplogFile` replay half: a dumped `local/oplog.rs.bson` is replayed against the live replica set and only the expected ops take effect.
- `test/legacy42/jstests/tool/dumprestore7.js` -> `TestOplogReplayFromLocalOplogRS` (`mongorestore/oplog_replset_test.go`) - the mongodump half: dumping `local.oplog.rs` with a `{ts: {$gt: ...}}` timestamp query captures exactly the ops after the checkpoint.

### CODE REVIEW

**Comment by autarch** [comment]
* **#1104** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1104?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1102** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1102?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1098** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1098?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1097** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1097?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1096** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1096?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1093** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1093?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1091** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1091?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1074** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1074?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1073** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1073?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1049** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1049?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1049?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1060** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1060?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1046** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1046?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1048](https://github.com/mongodb/mongo-tools/pull/1048) <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1048?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1041** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1041?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1047** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1047?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: mongorestore/oplog_replset_test.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+		_, err = coll.InsertOne(ctx, bson.D{{"_id", 100 + i}, {"batch", 2}})
+		require.NoError(t, err, "inserting second batch")
+	}
+
+	dumpDir := t.TempDir()
+	query := fmt.Sprintf(
+		`{"ts":{"$gt":{"$timestamp":{"t":%d,"i":%d}}},"ns":%q,"op":"i"}`,
+		checkpoint.T, checkpoint.I, ns,
+	)
+	require.NoError(t, runDump(t, baseToolOpts(t), dumpDir, func(md *mongodump.MongoDump) {
```
Comment (mmcclimon) [inline mongorestore/oplog_replset_test.go]: Should these really go in the `mongorestore` directory, since they're actually running dump + restore?

File: mongorestore/oplog_replset_test.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	// First batch: these ops precede the checkpoint and must NOT be replayed.
+	for i := range 5 {
+		_, err = coll.InsertOne(ctx, bson.D{{"_id", i}, {"batch", 1}})
+		require.NoError(t, err, "inserting first batch")
+	}
+
+	checkpoint := latestOplogTimestamp(t, client)
+
+	// Second batch: these ops follow the checkpoint and must be replayed.
+	const secondBatch = 7
```
Comment (mmcclimon) [inline mongorestore/oplog_replset_test.go]: (nit) This const is pretty weird, as is the choice of the number 7?

File: mongorestore/oplog_replset_test.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	require.NoError(t, result.Err, "replaying the dumped oplog succeeds")
+
+	count, err := coll.CountDocuments(ctx, bson.M{})
+	require.NoError(t, err, "counting replayed documents")
+	assert.EqualValues(
+		t, secondBatch, count,
+		"only the ops after the checkpoint are replayed",
+	)
+
+	batch1Count, err := coll.CountDocuments(ctx, bson.M{"batch": 1})
```
Comment (mmcclimon) [inline mongorestore/oplog_replset_test.go]: We usually use `bson.D` for all of these.

**Review by mmcclimon — APPROVED** [review]
Alright, though I'd probably move these to the dumprestore package and write them to match that style.

File: mongorestore/oplog_replset_test.go (hunk 2 above)
Comment (autarch) [reply mongorestore/oplog_replset_test.go]: Claude picked this semi-randomly, but it should differ from the first batch size, so we can be sure that it picks the right "side" of the oplog (with 7 entries _after_ the timestamp rather than 5 _before_).

File: mongorestore/oplog_replset_test.go (hunk 3 above)
Comment (autarch) [reply mongorestore/oplog_replset_test.go]: Fixed

File: mongorestore/oplog_replset_test.go (hunk 1 above)
Comment (autarch) [reply mongorestore/oplog_replset_test.go]: Moved

**Comment by autarch** [comment]
### Merge activity

* **Aug 12, 6:36 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/mongodb/mongo-tools/1049).
* **Aug 12, 6:42 PM UTC**: [Graphite](https://app.graphite.com/github/pr/mongodb/mongo-tools/1049) rebased this pull request as part of a merge.
* **Aug 12, 6:42 PM UTC**: @autarch merged this pull request with [Graphite](https://app.graphite.com/github/pr/mongodb/mongo-tools/1049).
